docs/design/storage_abstraction.md

```diff
@@ -0,0 +1,275 @@
+2. Resource Table Object is saved with Hash(Address + 0) as the key. The ResourceTable's key is the resource type, and the value is the Resource.
+3. Module Table Object is saved with Hash(Address + 1) as the key. The ModuleTable's key is the Module Name, and the value is the Module bytecode.
+
+In Rooch, each Table in Move corresponds to an SMT, and the Table Object mainly stores the root of the SMT. When loading data from a Table, the root of the SMT is first obtained through the Table Object ID, and the SMT is then initialized to obtain the value through the key. Tables can also be fields in Struct, with the serialized Table ID (TableHandle) stored in the Struct. In this way, Rooch's state tree is a hierarchy of nested SMTs.
+
+
+### StorageContext
+
+The StorageContext contains two fields: `TxContext`, which contains information related to the current transaction, and `ObjectStore`, which corresponds to the first layer of the state tree.
+
+```move
+module moveos_std::storage_context{
+    struct StorageContext {
+        tx_context: TxContext,
+        object_storage: ObjectStorage,
+    }
+}
+```
+
+Developers can define the StorageContext parameter in the `entry` method, and the MoveVM will automatically fill in the parameter.
+
+```move
+public entry function main(ctx: &mut StorageContext){
+    //function logic
+}
+```
+
+### RawTable
+
+In Rooch, a RawTable extension has been implemented for Move, which is an extension of the original Table and removes the constraints on Key and Value.
+
+The following are some of the key interfaces provided by RawTable in Move:
+
+```move
+module moveos_std::raw_table{
+    public(friend) fun add<K: copy + drop, V>(table_handle: address, key: K, val: V);
+
+    public(friend) fun borrow<K: copy + drop, V>(table_handle: address, key: K);
+
+    public(friend) fun borrow_mut<K: copy + drop, V>(table_handle: address, key: K): &mut V;
+
+    public(friend) fun remove<K: copy + drop, V>(table_handle: address, key: K): V;
+
+    public(friend) fun contains<K: copy + drop, V>(table_handle: address, key: K): bool;
+}
+```
+
+TableHandle is the ObjectID of the TableObject. The upper-level data structure of RawTable can directly operate on any type of K, V, like a raw-level KV storage interface. Each RowTable corresponds to a SMT in the StateDB. The ObjectID of the global ObjectStore is a special ObjectID with all bytes set to `0`.
+
+Currently, all methods in RawTable are `friend` methods, and developers cannot directly operate on them. They need to operate through the upper-level data structures that encapsulate RawTable, such as `Table<K,V>`, `TypeTable`, and `ObjectStorage`.
+
+### Rooch Object
+
+Objects in Rooch adopt a Box mode where they are equivalent to a unique ID box that can encapsulate a type `T` to create an Object.
+
+```move
+module moveos_std::object{
+    struct Object<T: key>{
+        id: ObjectID,
+        owner: address,
+        value: T,
```

Comment:
So, what could be the value here? Primitive types like integers and addresses? Can it be another object?

Reply:
`T` must have `key` ability, so it can not be Primitive types and also can not be another Object.

Reply:
so can an object own another object?

Reply:
No, we have ResourceAccount to do the same thing. And a struct can put an ObjectStorage as its field, so the Struct can hold many Objects.

```move
struct ChildrenObjects has key {
   children: ObjectStore,
}
```